src/matlab2tikz.m

```diff
@@ -3543,60 +3543,144 @@ function warnFacetedInterp(m2t)
 end
 % ==============================================================================
 function [m2t, str] = drawScatterPlot(m2t, h)
+    % DRAWSCATTERPLOT Draws a scatter plot
+    %
+    % A scatter plot is a plot containing only markers and where the
+    % size and/or color of each marker can be changed independently.
+    %
+    % References for TikZ code:
+    %  - http://tex.stackexchange.com/questions/197270/how-to-plot-scatter-points-using-pgfplots-with-color-defined-from-table-rgb-valu
+    %  - http://tex.stackexchange.com/questions/98646/multiple-different-meta-for-marker-color-and-marker-size
+    %
+    % See also: scatter
     str = '';
     if ~isVisible(h)
         return; % there is nothing to plot
     end
-
-    xData = get(h, 'XData');
-    yData = get(h, 'YData');
-    zData = get(h, 'ZData');
-    cData = get(h, 'CData');
-    sData = get(h, 'SizeData');
-
-    if isempty(cData) && strcmpi(getEnvironment(), 'Octave')
-        cData = get(h, 'MarkerEdgeColor');
+    
+    dataInfo   = getDataInfo(h, 'X','Y','Z','C','Size');
+    markerInfo = getMarkerInfo(m2t, h);
+    
+    if isempty(dataInfo.C) && strcmpi(getEnvironment(), 'Octave')
+        dataInfo.C = get(h, 'MarkerEdgeColor');
     end
 
-    markerInfo = getMarkerInfo(m2t, h);
     %TODO: check against getMarkerOptions() for duplicated code
 
-    constMarkerkSize = length(sData) == 1; % constant marker size
+    dataInfo.Size = tryToMakeScalar(dataInfo.Size, m2t.tol);
 
     % Rescale marker size (not definitive, follow discussion in #316)
     % Prescale marker size for octave
     if strcmpi(getEnvironment(), 'Octave')
-        sData = sData.^2/2;
+        dataInfo.Size = dataInfo.Size.^2/2;
     end
-    sData = translateMarkerSize(m2t, markerInfo.style, sqrt(sData)/2);
+    dataInfo.Size = translateMarkerSize(m2t, markerInfo.style, sqrt(dataInfo.Size)/2);
 
     drawOptions = opts_new();
-    if length(cData) == 3
-        [m2t, drawOptions] = getScatterOptsOneColor(m2t, h, drawOptions, ...
-                                                markerInfo, cData, sData, ...
-                                                constMarkerkSize);
-    elseif size(cData,2) == 3
-        drawOptions = getScatterOptsRGB(m2t, drawOptions);
+    
+    %% Determine if we are drawing an actual scatter plot
+    hasDifferentSizes  = numel(dataInfo.Size) ~= 1;
+    hasDifferentColors = numel(dataInfo.C)    ~= 3;
+    isaScatter         = hasDifferentSizes || hasDifferentColors;
+    if isaScatter
+        drawOptions = opts_add(drawOptions, 'scatter');
+    end
+    %TODO: we need to set the scatter source
+    drawOptions = opts_add(drawOptions, 'only marks');
+    drawOptions = opts_add(drawOptions, 'mark', markerInfo.tikz);
+
+    if length(dataInfo.C) == 3
```

Comment:
`length` and `numel` do something slightly different:
 - `length(x) == max(size(x))`
 - `numel(x) == prod(size(x))`

I had the impression it was something else (i.e. that `length` was `y = size(x); y(find(y~=1,1))`: the size of the first non-singleton dimension), so I think here having `length` is indeed dangerous.